pallets/qa-tools/src/lib.rs

```diff
@@ -0,0 +1,191 @@
+// are permitted provided that the following conditions are met:
+
+// Redistributions of source code must retain the above copyright notice, this list
+// of conditions and the following disclaimer.
+// Redistributions in binary form must reproduce the above copyright notice, this
+// list of conditions and the following disclaimer in the documentation and/or other
+// materials provided with the distribution.
+//
+// All advertising materials mentioning features or use of this software must display
+// the following acknowledgement: This product includes software developed by Polka Biome
+// Ltd., SORA, and Polkaswap.
+//
+// Neither the name of the Polka Biome Ltd. nor the names of its contributors may be used
+// to endorse or promote products derived from this software without specific prior written permission.
+
+// THIS SOFTWARE IS PROVIDED BY Polka Biome Ltd. AS IS AND ANY EXPRESS OR IMPLIED WARRANTIES,
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL Polka Biome Ltd. BE LIABLE FOR ANY
+// DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+// BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+// OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+pub use pallet::*;
+
+// private-net to make circular dependencies work
+#[cfg(all(test, feature = "private-net", feature = "ready-to-test"))] // order-book
+mod tests;
+pub mod weights;
+pub use weights::*;
+mod pallets;
+
+#[frame_support::pallet]
+pub mod pallet {
+    use super::*;
+    use common::{
+        AssetInfoProvider, AssetName, AssetSymbol, BalancePrecision, ContentSource, Description,
+    };
+    use frame_support::dispatch::DispatchErrorWithPostInfo;
+    use frame_support::traits::Get;
+    use frame_support::{dispatch::PostDispatchInfo, pallet_prelude::*};
+    use frame_system::pallet_prelude::*;
+    use order_book::{MomentOf, OrderBookId};
+    use sp_std::prelude::*;
+
+    #[pallet::pallet]
+    pub struct Pallet<T>(_);
+
+    #[pallet::config]
+    pub trait Config: frame_system::Config + order_book::Config + trading_pair::Config {
+        type WeightInfo: WeightInfo;
+        type OrderBookOrderLifespan: Get<MomentOf<Self>>;
+        type AssetInfoProvider: AssetInfoProvider<
+            Self::AssetId,
+            Self::AccountId,
+            AssetSymbol,
+            AssetName,
+            BalancePrecision,
+            ContentSource,
+            Description,
+        >;
+    }
+
+    #[pallet::error]
+    pub enum Error<T> {
+        /// Order book related errors
+        OrderBook(pallets::order_book_tools::Error),
```

Comment:
Does it work fine with polkadot.js? And how it is represented in metadata?

Reply:
I'll check it

Reply:
It works but doesn't show the underlying error :( 
<img width="444" alt="image" src="https://github.com/sora-xor/sora2-network/assets/8144358/261f10eb-6fb4-4921-a728-97e7c3a0a85c">
![image](https://github.com/sora-xor/sora2-network/assets/8144358/b9b9aa21-15ff-4b38-8a3d-527c55c1aae2)

Reply:
only `dispatchError.Module.error` field in polkadot.js distinguishes between underlying variants:
<img width="452" alt="image" src="https://github.com/sora-xor/sora2-network/assets/8144358/d7221d73-76d2-4868-ae47-d931037d65a5">
<img width="452" alt="image" src="https://github.com/sora-xor/sora2-network/assets/8144358/87257bc2-3c3a-4a3a-a62b-52ae7272f31f">

Reply:
So it seems like not the best solution in terms of polkadot.js user experience.

I wanted to somehow show that the error is related to order book pallet. The ways to do it that I see:

- Prefix error names with `OrderBook`, like `OrderBookUnknownOrderBook` _- seems kinda clumsy and inconvenient_
- Group them only in code - _not exposed externally (in particular to polkadot.js), so the users (testers in this case) should be expected to understand which pallet it refers to from context (might be ok considering that the error occurs right after the call)_
![image](https://github.com/sora-xor/sora2-network/assets/8144358/2a6d50a0-b1b7-490d-b83c-1a6afa18732b)
- Use existing error from `order_book` pallet (fits for this particular case, not sure about future versions of the pallet) - _should correctly display in polkadot.js, clearly separated in code, reuses existing errors (not sure if it's good or bad)_

Reply:
I lean towards prefixing in general, but reusing errors should be the best option in this particular case.

what do you think @vovac12 @wer1st ?

Reply:
@vovac12 can we have errors with fields? Such as `QaToolError(OrderBookError)` or `QaToolError(String)`?

Reply:
@bragov4ik you can use error names without prefix, group them in code and add detailed error description in doc comment

@wer1st no, dispatch error can only contain pallet index and error index

Reply:
Ok, seems like it's the best option in this case